### PR TITLE
Convert horizontal ellipsis in msgids

### PR DESCRIPTION
**Pull request by rubenswieringa**
**tl;dr:** angular-gettext-extract extracts `&hellip;` as a html-entity into .po-files, but browsers automatically convert `&hellip;` to `…`.

This means that your translations-file may have a msgid `"dots&hellip;"`, but the translate-directive will read `"dots…"` from a html-element and try to find that as a msgid (and find nothing).

To get around this we convert any &hellip; to … when compiling from a .po-file.

---

# The situation

Consider the following:

**page.html:**
```html
<p translate>Dots&hellip;</p>
```

**translations.po:**
```po
msgid "Dots&hellip;"
msgstr "Puntjes&hellip;"
```

Browsers automatically render `&hellip;` though.

**DOM:**
```html
<p translate>Dots…</p>
```

**Console:**
```javascript
p.innerHTML // -> "Dots…"
```

# The problem

So browsers automatically convert `&hellip;` to `…` (1 character). When the translate-directive tries to find the msgid `"Dots…"`, it will find nothing, because the translation is listed under `"Dots&hellip;"`.

There is no deterministic way to backtrack `…` to `&hellip;` in the DOM; a html-file may read `&hellip;` or `…` – in both cases the text-value in the DOM will read `…` (1 character)

In your .po-file, you will want the original text from the html-file though, so that you can translate *“Dots&hellip;”* to *“Puntjes&hellip;”*, instead of having to translate from *“Dots…”* (you’ll want to use the html-entity in your translated string).

# The solution

Bearing this in mind, it makes sense to convert all occurrences of `&hellip;` to `…` during compile. This means that msgids will still contain `&hellip;`, but the compiled translations-file will only contain `…` in msgids.

So in your .po-file you won’t notice anything, but when the translate-directive looks for `Dots…`, it wil actually find it.

### CODE REVIEW

**Comment by rubenv** [comment]
I guess this makes sense. Is this behavior consistent over browsers?

@gabegorelick @circlingthesun Any thoughts on this?

**Comment by rubenswieringa** [comment]
Yes, the behavior seems to be consistent across browsers:

**Test-case:** http://jsfiddle.net/8dtgpg6d/1/embedded/result

Excuse the lousy mobile screenshots – I made them straight out of browserstack (for convenience sake)

# Windows

* **IE9:** <br/>ie9
* **IE10:** <br/>ie10
* **IE11:** <br/>ie11
* **IE12 (Edge):** <br/>ie_edge12


# OSX

* **Chrome 47 (OSX):** chrome_osx
* **Safari 9.0.1 (OSX):** safari_osx
* **Firefox 42 (OSX):** firefox_osx


# iOS

* **iOS7:** <br/>safari_ios7
* **iOS8:** <br/>safari_ios8
* **iOS9:** <br/>safari_ios9


# Android

* **Android4.0 (stock)** <br/>android4 0_galaxynote
* **Android4.1 (stock)** <br/>android4 1_galaxynote
* **Android4.2 (stock)** <br/>android4 2_nexus
* **Android4.3 (stock)** <br/>android4 3_galaxynote
* **Android4.4 (stock)** <br/>android4 4_galaxynote
* **Android5.0 (stock)** <br/>android5 0_googlenexus
* **Android5.1 (stock)** <br/>android5 1_googlenexus




# Windows Phone

* <br/>windows_phone

**Comment by rubenv** [comment]
Wow, that's a fantastically thorough piece of research! Thanks for taking the time to do this.

My concerns are gone, I see no further reason not to merge this. We can always revert if it causes trouble (doubt it).

Excellent contribution, thanks a lot. Release coming!

**Comment by rubenv** [comment]
Released to NPM as `2.1.10`. Will also release a new `grunt-angular-gettext` that depends on this version.

**Comment by rubenswieringa** [comment]
Sweet! Thanks for the quick response!

**Comment by rubenv** [comment]
All the credit goes to you for the hard work!
